File: tools/bag_to_pcd.cpp
// ros2 run 一次即可将 /livox/lidar 的点云攒成一帧保存为 PCD
// 用法：先启动雷达，再运行本节点，等几秒后 Ctrl+C
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <pcl_conversions/pcl_conversions.h>
#include <pcl/io/pcd_io.h>
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>

class BagToPcd : public rclcpp::Node {
public:
    BagToPcd() : Node("bag_to_pcd") {
        // 从参数接收录制时长（秒）和输出路径
        this->declare_parameter("duration", 10.0);
        this->declare_parameter("output", "config/field.pcd");
        duration_ = this->get_parameter("duration").as_double();
        output_ = this->get_parameter("output").as_string();

        cloud_.reset(new pcl::PointCloud<pcl::PointXYZ>);
        sub_ = this->create_subscription<sensor_msgs::msg::PointCloud2>(
            "/livox/lidar", rclcpp::SensorDataQoS(),
            std::bind(&BagToPcd::callback, this, std::placeholders::_1));

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(static_cast<int>(duration_ * 1000)),
            std::bind(&BagToPcd::finish, this));

        RCLCPP_INFO(this->get_logger(), "Recording for %.1f seconds...", duration_);
    }

private:
    void callback(const sensor_msgs::msg::PointCloud2::SharedPtr msg) {
        pcl::PointCloud<pcl::PointXYZ> temp;
        pcl::fromROSMsg(*msg, temp);
        *cloud_ += temp;
        RCLCPP_INFO_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
            "Accumulated %zu points", cloud_->size());
    }

    void finish() {
        cloud_->header.frame_id = "livox_frame";
        pcl::io::savePCDFileBinary(output_, *cloud_);
        RCLCPP_INFO(this->get_logger(),
            "Saved %zu points to %s", cloud_->size(), output_.c_str());
        rclcpp::shutdown();
    }

    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr sub_;
    rclcpp::TimerBase::SharedPtr timer_;
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_;
    double duration_;
    std::string output_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<BagToPcd>());
    return 0;
}

File: tools/bag_to_pcd.py
#!/usr/bin/env python3
"""订阅 /livox/lidar 攒 N 秒点云，保存为 PCD 文件。"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import PointCloud2
import sensor_msgs_py.point_cloud2 as pc2


class BagToPcd(Node):
    def __init__(self):
        super().__init__('bag_to_pcd')
        self.declare_parameter('duration', 10)
        self.declare_parameter('output', 'config/field.pcd')
        duration = self.get_parameter('duration').value
        self.output = self.get_parameter('output').value

        self.points = []  # [(x, y, z), ...]

        qos = QoSProfile(depth=10, reliability=ReliabilityPolicy.BEST_EFFORT)
        self.sub = self.create_subscription(
            PointCloud2, '/livox/lidar', self.callback, qos)

        self.timer = self.create_timer(duration, self.finish)
        self.get_logger().info(f'Recording for {duration}s -> {self.output}')

    def callback(self, msg):
        gen = pc2.read_points(msg, field_names=('x', 'y', 'z'), skip_nans=True)
        for p in gen:
            self.points.append(p)
        self.get_logger().info(f'Accumulated {len(self.points)} points', throttle_duration_sec=2)

    def finish(self):
        self._save_pcd(self.points, self.output)
        self.get_logger().info(f'Saved {len(self.points)} points to {self.output}')
        rclpy.shutdown()

    @staticmethod
    def _save_pcd(points, path):
        with open(path, 'w') as f:
            f.write(
                '# .PCD v0.7 - Point Cloud Data file format\n'
                'VERSION 0.7\n'
                'FIELDS x y z\n'
                'SIZE 4 4 4\n'
                'TYPE F F F\n'
                'COUNT 1 1 1\n'
                f'WIDTH {len(points)}\n'
                'HEIGHT 1\n'
                'VIEWPOINT 0 0 0 1 0 0 0\n'
                f'POINTS {len(points)}\n'
                'DATA ascii\n'
            )
            for p in points:
                f.write(f'{p[0]} {p[1]} {p[2]}\n')


def main():
    rclpy.init()
    node = BagToPcd()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.finish()


if __name__ == '__main__':
    main()
